homeassistant/components/nibe_heatpump/__init__.py

```diff
@@ -0,0 +1,242 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Nibe Heat Pump from a config entry."""
+
+    heatpump = HeatPump(Model[entry.data[CONF_MODEL]])
+    heatpump.word_swap = entry.data[CONF_WORD_SWAP]
+    heatpump.initialize()
+
+    connection_type = entry.data[CONF_CONNECTION_TYPE]
+
+    if connection_type == CONF_CONNECTION_TYPE_NIBEGW:
+        connection = NibeGW(
+            heatpump,
+            entry.data[CONF_IP_ADDRESS],
+            entry.data[CONF_REMOTE_READ_PORT],
+            entry.data[CONF_REMOTE_WRITE_PORT],
+            listening_port=entry.data[CONF_LISTENING_PORT],
+        )
+    else:
+        raise HomeAssistantError(f"Connection type {connection_type} is not supported.")
+
+    await connection.start()
+    try:
+        coordinator = Coordinator(hass, heatpump, connection)
+
+        data = hass.data.setdefault(DOMAIN, {})
+        data[entry.entry_id] = coordinator
+        await coordinator.async_config_entry_first_refresh()
+
+        reg = dr.async_get(hass)
+        reg.async_get_or_create(
+            config_entry_id=entry.entry_id,
+            identifiers={(DOMAIN, entry.unique_id or entry.entry_id)},
+            manufacturer="NIBE Energy Systems",
+            model=heatpump.model.name,
+            name=heatpump.model.name,
+        )
+
+        await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    except Exception:
+        await connection.stop()
+        raise
+
+    # Trigger a refresh again now that all platforms have registered
+    hass.async_add_job(coordinator.async_refresh)
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        coordinator: Coordinator = hass.data[DOMAIN].pop(entry.entry_id)
+        await coordinator.connection.stop()
+
+        # sleep a bit to make sure sockets are closed
+        await asyncio.sleep(5)
+
+    return unload_ok
+
+
+class Coordinator(DataUpdateCoordinator[dict[int, Coil]]):
+    """Update coordinator for nibe heat pumps."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        heatpump: HeatPump,
+        connection: Connection,
+    ) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass, _LOGGER, name="Nibe Heat Pump", update_interval=timedelta(seconds=60)
+        )
+
+        self.connection = connection
+        self.heatpump = heatpump
+
+    @property
+    def coils(self) -> dict[str, Coil]:
+        """Return the full coil database."""
+        return self.heatpump._address_to_coil  # pylint: disable=protected-access
+
+    @property
+    def unique_id(self) -> str:
+        """Return unique id for this coordinator."""
+        assert self.config_entry
```

Comment:
Instead of asserting this in several places, why not just declare that you expect it to be a part of this class above `__init__?

```python
config_entry: ConfigEntry
```

Reply:
It's part of the standard DataUpdateCoordinator, but i assume we could assume it will always set it up for this coordinator.